src/main/java/org/osc/manager/rest/server/api/ManagerConnectorApis.java

```diff
@@ -0,0 +1,234 @@
+                if (result.isEmpty()) {
+                    throw new Exception("Manager Connector Entity does not already exists..");
+                    //TODO - to add RETURN 404 error:Sudhir
+                }
+                result.get(0).setName(entity.getName());
+                em.persist(result.get(0));
+                return entity;
+
+            }
+        });
+    }
+
+    /**
+     * Delete MC for a given ID
+     *
+     * @return - deleted MC
+     */
+    @Path("/{applianceManagerConnectorId}")
+    @DELETE
+    public void deleteManagerConnector(@PathParam("applianceManagerConnectorId") Long amcId,
+            ApplianceManagerConnectorEntity entity) {
+
+        logger.info("Deleting MC Entity Id...:" + amcId);
+
+        this.txControl.required(new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+
+                CriteriaBuilder criteriaBuilder = ManagerConnectorApis.this.em.getCriteriaBuilder();
+
+                CriteriaQuery<ApplianceManagerConnectorEntity> query = criteriaBuilder
+                        .createQuery(ApplianceManagerConnectorEntity.class);
+                Root<ApplianceManagerConnectorEntity> r = query.from(ApplianceManagerConnectorEntity.class);
+                query.select(r).where(criteriaBuilder.and(criteriaBuilder.equal(r.get("id"), amcId)));
+
+                List<ApplianceManagerConnectorEntity> result = em.createQuery(query).getResultList();
+                if (result.isEmpty()) {
+                    return null;
+                }
+                em.remove(result.get(0));
+                return null;
+            }
+        });
+    }
+
+    /**
+     * Lists the MC for a given ID
+     *
+     * @return - MC
+     */
+    @Path("/{applianceManagerConnectorId}")
+    @GET
+    public ApplianceManagerConnectorEntity getManagerConnector(@PathParam("applianceManagerConnectorId") Long amcId) {
+
+        logger.info("Query MC for Id...:" + amcId);
+
+        return this.txControl.required(new Callable<ApplianceManagerConnectorEntity>() {
+            @Override
+            public ApplianceManagerConnectorEntity call() throws Exception {
+                CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();
+                CriteriaQuery<ApplianceManagerConnectorEntity> query = criteriaBuilder
+                        .createQuery(ApplianceManagerConnectorEntity.class);
+                Root<ApplianceManagerConnectorEntity> r = query.from(ApplianceManagerConnectorEntity.class);
+                query.select(r).where(criteriaBuilder.and(criteriaBuilder.equal(r.get("id"), amcId)));
+
+                List<ApplianceManagerConnectorEntity> result = em.createQuery(query).getResultList();
+                if (result.isEmpty()) {
+                    return null;
+                }
+                ApplianceManagerConnectorEntity mc = new ApplianceManagerConnectorEntity();
+                mc.setName(result.get(0).getName());
```

Comment:
@sudhirappaji I believe this is happening because the API is returning before the transaction finishes. Look at the exising IsmDeviceApi.java. You may need to update your code throughout with either one of these and hopefully that will take care of the issue while letting you return the element. See:

```java
return txControl.required(new Callable<DeviceMember>() {
            @Override
            public DeviceMember call() throws Exception {
                VSSDevice device = em.find(VSSDevice.class, vs.getId());
                
                DeviceMember dm = new DeviceMember();
                dm.setParent(device);
                dm.setName(name);
                em.persist(dm);
                return dm;
            }
        }).getId().toString();  // creates include the retrieval of this id this way which must be forcing the conclusion
```
or this:
```java
txControl.supports(new Callable<ManagerDeviceMemberElement>() { // supports is being used for gets/retrieval
            
            @Override
            public ManagerDeviceMemberElement call() throws Exception {
                DeviceMember result = em.find(DeviceMember.class, Long.valueOf(id));
                return result == null ? null : new MemberDevice(result.getId(), result.getName());
            }
        });
```
Both of which are slightly different than your code. Can you look at these approaches and see when to use .supports and when to use .required with

Reply:
@dmitryintel fyi

Reply:
I see the em.find method is similar to my solution and it is lot clearner. But looks like em.find method does not give fine control over the query and existing device code juggles between find and query calls.
I tried to check em.find and returning result directly, it throws up same issue. one needs to copy result and then return. I will wait for your response before I implement for the rest of the APIs'.
Also I cannot create the constructor in the domain or policyentity class, as hibernate triggers runtime exception saying constructor should not be defined.

Reply:
I will also explore other methods to synchronize the query path... and will keep you posted.

Reply:
continuation of the above investigation for the find, I looked at the wait and notify methods of txcontrol. placement of wait should be done before find method gets triggered from the thread context. which is not feasible.
even if it is feasible, it becomes complicated when the notifier should be called especially if exceptions gets triggered, notify is not invoked the entire chain is blocked. so not recommended.
in my view we should leave with copy for now for the get accesses either via em.find or query methods depending complexity of query. let me know what you think?

Reply:
Update ==> for txcontrol.required approach as above works. one needs to return always String type in our example. so update method shall change to return ID rather then the domain or policy entity. The snippet of the code is attached below:
public String updateDomain(@PathParam("domainid") Long domainid, DomainEntity entity) {

        logger.info("Updating Domain Entity ID...:" + domainid);

        return this.txControl.required(new Callable<DomainEntity>() {

            @Override
            public DomainEntity call() throws Exception {

                DomainEntity result = em.find(DomainEntity.class, domainid);
                if (result == null) {
                    throw new Exception("Domain Entity does not exists...");
                    //TODO - to add RETURN 404 error:Sudhir
                }
                result.setName(entity.getName());
                em.persist(result);
                return result;
            }
        }).getId();
    }

Reply:
further update and hopefully final one... I tried to return the result directly from getDomain after adding jsonignore annotation for policy and appliancemanagerconnector. It works with side effects of adding the jsonignore annotation. this annotation does not allow this entities to be accessed at all nor persist. so this is not workable solution. 
The real cause of this problem appear to be when we return the 'result', it looks like framework buffer is becoming full and rest layer continues to send the response without data. so copying only a subset of this data and return works.
now in my view we have the overall picture about the issue, can we now copy and  return?

Reply:
i've talked to sudhir and i mentioned that objects with lazy loaded entities(domain lazy loads policies) jersey will have issues since it will try to access the policy entity after the DB connection is closed....i would try with all eager loading and see if the issue persists

Reply:
Yes it works with EAGER fetch.type and will continue to do more testing and will update with any new findings...